feat: LEAP-1439: Display link in existing comment and allow linking

Gondragos opened the pull request:
This PR provides an ability to link an existing comment to a region and display linked region.

Other changes:
- restoring some css variables in LSF https://github.com/HumanSignal/label-studio/pull/6404/files#diff-807fd357841e1fae140465865a611c4ba4d24b27dae837d2f4193d82e60ef78eR6-R47

### PR fulfills these requirements
- [x] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)



#### Change has impacts in these area(s)
- [ ] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [x] Frontend



### Describe the reason for change
It's a part of implementing linking comment to region functionality.



#### What feature flags were used to cover this change?
`fflag_feat_all_leap_1430_per_field_comments_100924_short`



### Does this PR introduce a breaking change?
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [x] No
- [ ] Not sure (briefly explain the situation below)



### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [ ] integration
- [ ] unit



### Which logical domain(s) does this change affect?
`Comments`

Gondragos committed:
Add Linking Modes for Annotations

Introduce `CommentMode` and `RelationMode` to handle linking annotations with regions. Also, refactor parts of the code to unify the method of linking actions under `LinkingModes` and replace deprecated `relationMode` checks.

Gondragos committed:
Add integration tests for linking modes to comments

Gondragos committed:
Refactor CommentMode and RelationMode to use volatile state.

Moved the `comment` and `firstRegion` variables to volatile state within the MobX-state-tree models for `CommentMode` and `RelationMode`. This change provides access to payload information of modes in case we need it.

Commit:
Fix warning message

Co-authored-by: hlomzik <hlomzik@gmail.com>

Commit:
Fix comment about deprecated method

Co-authored-by: hlomzik <hlomzik@gmail.com>

Commit:
Fix comment about deprecated method

Co-authored-by: hlomzik <hlomzik@gmail.com>

Gondragos committed:
Rename misnamed 'to_coments' to 'to_comments'

Gondragos committed:
Change cursor constants for linking mode

Updated all instances of `Constants.RELATION_MODE_CURSOR` to `Constants.LINKING_MODE_CURSOR` across various files. This ensures consistency in naming and improves code readability for linking mode features.

Gondragos committed:
Refactor LinkingModes model

Gondragos committed:
Fix linkingModes initing

Gondragos committed:
Update comment handling and linking UI

Refactored comments feature to support per-field comments through feature flags, added linking functionality between comments and regions, and updated UI with new icons and styles.

Gondragos committed:
Refactor comment form component structure

Reorganize the comment form components by separating old and new versions. Adjust CSS class names to avoid conflicts and import paths for clarity.

Gondragos committed:
Fix typo in CommentForm.scss

Gondragos committed:
Provide linking region functionality for existing comment

Gondragos committed:
Restore some missed colors

Gondragos committed:
Fix types

Gondragos committed:
Merge branch 'develop' into 'fb-leap-1440/link-new-comment'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10942180936

Gondragos committed:
Fix FF_OUTLINER_OPTIM integration

Gondragos committed:
Refactor display of region labels

Replace `displayRegionLabel` with `RegionLabel` component in OutlinerTree and LinkState components.

Gondragos committed:
Remove unused IconWarning and Tooltip imports

Gondragos committed:
stylelint fixes

Gondragos committed:
biome fixes

Gondragos committed:
Missed file

Gondragos commented:
/git merge fb-leap-1440/link-new-comment
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10944353621)
> Successfully merged:  create mode 100644 web/libs/editor/tests/integration/e2e/control_tags/classification/textarea.cy.ts

Gondragos committed:
Merge branch 'fb-leap-1440/link-new-comment' into 'fb-leap-1439/link-comment'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10944353621

Commit:
Add blank lines for readability

Co-authored-by: bmartel <brandonmartel@gmail.com>

Commit:
Add blank line at eof

Co-authored-by: bmartel <brandonmartel@gmail.com>

Commit:
Add blank lines for readability

Co-authored-by: bmartel <brandonmartel@gmail.com>

Gondragos committed:
Remove block variable from scss

Gondragos committed:
Clear commented code

Gondragos committed:
stylelint fixes

Gondragos committed:
Adjust opacity for resolved comments

Gondragos committed:
Remove SCSS block variable

Gondragos committed:
Merge branch 'fb-leap-1440/link-new-comment' into fb-leap-1439/link-comment

# Conflicts:
#	web/libs/editor/src/components/Comments/Comment/CommentForm.scss
#	web/libs/editor/src/components/Comments/Comment/LinkState.scss

Gondragos committed:
Rename comment form class to avoid intersections again

yyassi-heartex commented on web/libs/editor/public/styles/main.css:
```diff
@@ -2,6 +2,49 @@ body {
   margin: 0;
   padding: 0;
+
+  --sand_900: #12110D;
```
these colors are defined in web/apps/labelstudio/src/themes/default/colors.scss - are comments not being pulled into a spot that pulls from that?

Gondragos replied:
`web/apps/labelstudio` is LSO, but we have LSF standalone and it's broken a bit. There is also some problem at least with `border` in some elements.

Gondragos replied:
For example:
<img width="332" alt="image" src="https://github.com/user-attachments/assets/e40224d0-f63c-4ab5-b236-36fd0ebdbfa9">
<img width="415" alt="image" src="https://github.com/user-attachments/assets/c1628f8c-abe7-4122-9ea2-53606227b950">
instead of
<img width="343" alt="image" src="https://github.com/user-attachments/assets/c139f56c-6bb9-4a02-9ee6-12b4de53334e">
<img width="328" alt="image" src="https://github.com/user-attachments/assets/c242fdf3-7400-4b47-8e73-ffc061a18ff7">

yyassi-heartex commented on web/libs/editor/src/components/Comments/Comment/CommentItem.scss:
```diff
@@ -0,0 +1,142 @@
+}
+
+.comment-item {
+  --comment-item-padding: 8px;
+  --comment-item-spacing: 6px;
+  --comment-item-content-font-size: 14px;
+  --comment-item-content-line-height: 20px;
+  --comment-item-content-letter-spacing: 0.25px;
+  --comment-item-avatar-size: 24px;
+  --comment-item-name-color: var(--sand_900);
+  --comment-item-content-color: var(--sand_900);
+  --comment-item-content-color-resolved: var(--sand_500);
+  --comment-item-icon-color-resolved: var(--success_color);
+  --comment-item-region-opacity: 1;
```
❤️ 
this is great!

Gondragos replied:
Oh, it mostly was already there before 😊 https://github.com/HumanSignal/label-studio/blob/develop/web/libs/editor/src/components/Comments/CommentItem.scss#L12-L21

yyassi-heartex commented on web/libs/editor/src/components/Comments/Comments.tsx:
```diff
@@ -1,14 +1,17 @@
 import "./Comments.scss";
+
 const CommentForm = isFF(FF_PER_FIELD_COMMENTS) ? CommentsFormNew : CommentFormOld;
+const CommentsList = isFF(FF_PER_FIELD_COMMENTS) ? CommentsListNew : CommentsListOld;
```
minor nitpick - since `isFF(FF_PER_FIELD_COMMENTS)` is reused - lets check it once and reuse the calculated value

yyassi-heartex reviewed (approved):
looks good over all - i have a question and a minor nit-pick other wise ✅

Gondragos committed:
Refactor feature flag checks in Comments component

Refactor the feature flag checks by introducing the `isPerFieldComments` variable.

Gondragos commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/11015174443)
> Successfully merged:  23 files changed, 702 insertions(+), 71 deletions(-)

Gondragos committed:
Merge branch 'develop' into 'fb-leap-1439/link-comment'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/11015174443

Gondragos committed:
Fix bad merge

Gondragos committed:
stylelint fixes

hlomzik commented on web/libs/editor/src/stores/Comment/Comment.js:
```diff
@@ -115,6 +113,32 @@ export const Comment = CommentBase.named("Comment")
+          id: self.id,
+          ...props,
+        });
+        if (result.error) return;
```
we should also reset `isUpdating` here

hlomzik commented on web/libs/editor/src/stores/Comment/Comment.js:
```diff
@@ -115,6 +113,32 @@ export const Comment = CommentBase.named("Comment")
+      const regionRef = {
+        regionId: region.cleanId,
+      };
+      self.update(snakeizeKeys({ regionRef }));
```
things like this should be on a same level to incapsulate knowledge. move this transformation to the `update()` which also uses reversed transform.

Gondragos committed:
Refactor comment update mechanism

Improve the comment update process by incorporating `snakeizeKeys` for `props`. Properly handle update error state by setting `isUpdating` to false when errors occur. Simplify `regionRef` processing by removing unnecessary `snakeizeKeys` calls on upper level.

Gondragos commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/11033272280)
> Successfully merged:  create mode 100644 web/dist/apps/labelstudio/link-to.539fa385668368f5e3c242aa876ae6d3.svg

Gondragos committed:
Merge branch 'develop' into 'fb-leap-1439/link-comment'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/11033272280

hlomzik commented on web/libs/editor/src/components/Comments/Comment/CommentsList.tsx:
```diff
@@ -0,0 +1,48 @@
+    },
+    [commentStore],
+  );
+  const globalLinking = commentStore.annotation && commentStore.annotation.linkingMode === LINK_COMMENT_MODE;
```
```suggestion
  const globalLinking = commentStore.annotation?.linkingMode === LINK_COMMENT_MODE;
```

hlomzik commented on web/libs/editor/src/components/Comments/Comment/CommentItem.tsx:
```diff
@@ -0,0 +1,211 @@
+                      >
+                        {isEditMode ? "Cancel edit" : "Edit"}
+                      </Menu.Item>
+                      <Menu.Item onClick={toggleLink}>{regionRef?.region ? "Unlink" : "Link to ..."}</Menu.Item>
```
```suggestion
                      <Menu.Item onClick={toggleLink}>{regionRef?.region ? "Unlink" : "Link to..."}</Menu.Item>
```
dialog items in menus should have ellipsis right after the text

Gondragos committed:
Fix new comment management in CommentForm and CommentStore.

Updated CommentForm to use commentInProgress instead of currentComment for improved clarity. Also, included annotationStore in CommentBase creation to ensure consistent data access.

Gondragos committed:
Merge branch 'fb-leap-1439/link-comment' of ssh://github.com/heartexlabs/label-studio into fb-leap-1439/link-comment

Commit:
Apply suggestions from code review

Co-authored-by: hlomzik <hlomzik@gmail.com>

Gondragos committed:
biomejs fixes

Gondragos committed:
Merge branch 'fb-leap-1439/link-comment' of ssh://github.com/heartexlabs/label-studio into fb-leap-1439/link-comment

Gondragos commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/11043237251)
> Successfully merged:  create mode 100644 web/libs/editor/src/regions/TextAreaRegion/TextAreaRegion.scss

Gondragos committed:
Merge branch 'develop' into 'fb-leap-1439/link-comment'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/11043237251
